plugins/module_utils/attachment.py

```diff
@@ -0,0 +1,192 @@
+            )
+
+        if must_exist and not records:
+            raise errors.ServiceNowError(
+                "No attachments match the {0} query.".format(query)
+            )
+
+        return records[0] if records else None
+
+    def create_record(self, query, data, check_mode, mime_type):
+        if check_mode:
+            # Approximate the result using the payload and data.
+            return query
+        return self.client.request(
+            "POST",
+            _path("file"),
+            query=(query or {}),
+            headers={"Accept": "application/json", "Content-type": mime_type},
+            bytes=data,
+        ).json["result"]
+
+    def upload_record(self, table, table_sys_id, metadata, check_mode=False):
+        # Table and table_sys_id parameters uniquely identify the record we will attach a file to.
+        #
+        # "metadata" is a dict with a mandatory key "path" and optional "name", "type" and "encryption_context".
+        # These properties can be read directly from the yaml attachment list and can be set by the user.
+        query = dict(table_name=table, table_sys_id=table_sys_id, **metadata)
+        if "name" in query:
+            query["file_name"] = query.pop("name")
+        if "type" in query:
+            query["content_type"] = query.pop("type")
+        try:
+            with open(metadata["path"], "rb") as file_obj:
+                data = file_obj.read()
+        except (IOError, OSError):
+            raise errors.ServiceNowError("Cannot open {0}".format(query["path"]))
+        return self.create_record(query, data, check_mode, query["content_type"])
+
+    def upload_records(self, table, table_sys_id, metadata_dict, check_mode):
+        return [
+            self.upload_record(
+                table, table_sys_id, dict(name=name, **metadata), check_mode
+            )
+            for name, metadata in metadata_dict.items()
+        ]
+
+    def delete_record(self, record, check_mode):
+        if not check_mode:
+            self.client.delete(_path(record["sys_id"]))
+
+    def delete_attached_records(self, table, table_sys_id, check_mode):
+        return [
+            self.delete_record(record, check_mode)
+            for record in self.list_records(
+                dict(
+                    table_name=table,
+                    table_sys_id=table_sys_id,
+                )
+            )
+        ]
+
+    def update_records(
+        self, table, table_sys_id, metadata_dict, records, check_mode=False
+    ):
+        mapped_records = dict()
+        for record in records:
+            mapped_records[record["file_name"]] = record
+
+        for name, metadata in metadata_dict.items():
+            record = mapped_records.get(name, None)
+
+            if record is not None and record["hash"] == metadata["hash"]:
+                mapped_records[name] = dict(
+                    {"changed": False, "msg": "Skipped. Hash matches remote."}, **record
+                )
```

Comment:
We only return changes and other stuff at the top level. Here we should just make sure mapped records contain the state on the remote at the end of the update.

```suggestion
                continue
```

Reply:
I fixed this a bit differently, but it does the same job.